{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "168719a9",
   "metadata": {},
   "source": [
    "Install Bioconductor and Annotatehub to download the annotation files"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fab88899",
   "metadata": {},
   "source": [
    "I faced some problems installing the tools because R was old. makesure your R version is 4.2. I used conda to install the R but you can update R anyway"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "ff2d16e0",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cloud.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.15 (BiocManager 1.30.18), R 4.2.0 (2022-04-22)\n",
      "\n",
      "Old packages: 'aweek', 'brew', 'callr', 'cli', 'coloc', 'covr', 'crayon',\n",
      "  'curl', 'desc', 'dplyr', 'DT', 'epiR', 'evaluate', 'gargle', 'gert', 'gh',\n",
      "  'gitcreds', 'gmp', 'gtable', 'httpuv', 'httr', 'hunspell', 'incidence',\n",
      "  'irlba', 'isoband', 'jsonlite', 'knitr', 'lifecycle', 'lintr',\n",
      "  'MatrixModels', 'MendelianRandomization', 'openssl', 'pillar', 'prevalence',\n",
      "  'purrr', 'ragg', 'RcppArmadillo', 'rhub', 'rlang', 'rmarkdown', 'rstudioapi',\n",
      "  'rversions', 'rvest', 'scales', 'stringr', 'susieR', 'testthat', 'tidyr',\n",
      "  'tidyselect', 'tinytex', 'vctrs', 'viridisLite', 'xfun', 'zip', 'cluster',\n",
      "  'foreign', 'MASS', 'Matrix', 'nlme', 'nnet', 'survival'\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cloud.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.15 (BiocManager 1.30.18), R 4.2.0 (2022-04-22)\n",
      "\n",
      "Old packages: 'aweek', 'brew', 'callr', 'cli', 'coloc', 'covr', 'crayon',\n",
      "  'curl', 'desc', 'dplyr', 'DT', 'epiR', 'evaluate', 'gargle', 'gert', 'gh',\n",
      "  'gitcreds', 'gmp', 'gtable', 'httpuv', 'httr', 'hunspell', 'incidence',\n",
      "  'irlba', 'isoband', 'jsonlite', 'knitr', 'lifecycle', 'lintr',\n",
      "  'MatrixModels', 'MendelianRandomization', 'openssl', 'pillar', 'prevalence',\n",
      "  'purrr', 'ragg', 'RcppArmadillo', 'rhub', 'rlang', 'rmarkdown', 'rstudioapi',\n",
      "  'rversions', 'rvest', 'scales', 'stringr', 'susieR', 'testthat', 'tidyr',\n",
      "  'tidyselect', 'tinytex', 'vctrs', 'viridisLite', 'xfun', 'zip', 'cluster',\n",
      "  'foreign', 'MASS', 'Matrix', 'nlme', 'nnet', 'survival'\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cloud.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.15 (BiocManager 1.30.18), R 4.2.0 (2022-04-22)\n",
      "\n",
      "Installing package(s) 'AnnotationHub'\n",
      "\n",
      "also installing the dependencies ‘bitops’, ‘RCurl’, ‘GenomeInfoDbData’, ‘XVector’, ‘GenomeInfoDb’, ‘bit’, ‘png’, ‘Biostrings’, ‘dbplyr’, ‘DBI’, ‘filelock’, ‘bit64’, ‘blob’, ‘plogr’, ‘Biobase’, ‘IRanges’, ‘KEGGREST’, ‘BiocFileCache’, ‘RSQLite’, ‘AnnotationDbi’, ‘S4Vectors’, ‘interactiveDisplayBase’\n",
      "\n",
      "\n",
      "Old packages: 'aweek', 'brew', 'callr', 'cli', 'coloc', 'covr', 'crayon',\n",
      "  'curl', 'desc', 'dplyr', 'DT', 'epiR', 'evaluate', 'gargle', 'gert', 'gh',\n",
      "  'gitcreds', 'gmp', 'gtable', 'httpuv', 'httr', 'hunspell', 'incidence',\n",
      "  'irlba', 'isoband', 'jsonlite', 'knitr', 'lifecycle', 'lintr',\n",
      "  'MatrixModels', 'MendelianRandomization', 'openssl', 'pillar', 'prevalence',\n",
      "  'purrr', 'ragg', 'RcppArmadillo', 'rhub', 'rlang', 'rmarkdown', 'rstudioapi',\n",
      "  'rversions', 'rvest', 'scales', 'stringr', 'susieR', 'testthat', 'tidyr',\n",
      "  'tidyselect', 'tinytex', 'vctrs', 'viridisLite', 'xfun', 'zip', 'cluster',\n",
      "  'foreign', 'MASS', 'Matrix', 'nlme', 'nnet', 'survival'\n",
      "\n",
      "'getOption(\"repos\")' replaces Bioconductor standard repositories, see\n",
      "'?repositories' for details\n",
      "\n",
      "replacement repositories:\n",
      "    CRAN: https://cloud.r-project.org\n",
      "\n",
      "\n",
      "Bioconductor version 3.15 (BiocManager 1.30.18), R 4.2.0 (2022-04-22)\n",
      "\n",
      "Installing package(s) 'StructuralVariantAnnotation'\n",
      "\n",
      "also installing the dependencies ‘formatR’, ‘lambda.r’, ‘futile.options’, ‘futile.logger’, ‘snow’, ‘BH’, ‘hms’, ‘BiocParallel’, ‘progress’, ‘XML’, ‘Rsamtools’, ‘GenomicAlignments’, ‘BiocIO’, ‘restfulr’, ‘MatrixGenerics’, ‘BSgenome’, ‘Rhtslib’, ‘biomaRt’, ‘DelayedArray’, ‘GenomicRanges’, ‘rtracklayer’, ‘VariantAnnotation’, ‘GenomicFeatures’, ‘SummarizedExperiment’\n",
      "\n",
      "\n",
      "Old packages: 'aweek', 'brew', 'callr', 'cli', 'coloc', 'covr', 'crayon',\n",
      "  'curl', 'desc', 'dplyr', 'DT', 'epiR', 'evaluate', 'gargle', 'gert', 'gh',\n",
      "  'gitcreds', 'gmp', 'gtable', 'httpuv', 'httr', 'hunspell', 'incidence',\n",
      "  'irlba', 'isoband', 'jsonlite', 'knitr', 'lifecycle', 'lintr',\n",
      "  'MatrixModels', 'MendelianRandomization', 'openssl', 'pillar', 'prevalence',\n",
      "  'purrr', 'ragg', 'RcppArmadillo', 'rhub', 'rlang', 'rmarkdown', 'rstudioapi',\n",
      "  'rversions', 'rvest', 'scales', 'stringr', 'susieR', 'testthat', 'tidyr',\n",
      "  'tidyselect', 'tinytex', 'vctrs', 'viridisLite', 'xfun', 'zip', 'cluster',\n",
      "  'foreign', 'MASS', 'Matrix', 'nlme', 'nnet', 'survival'\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: rtracklayer\n",
      "\n",
      "Loading required package: VariantAnnotation\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "Loading required package: Rsamtools\n",
      "\n",
      "Loading required package: Biostrings\n",
      "\n",
      "Loading required package: XVector\n",
      "\n",
      "\n",
      "Attaching package: ‘Biostrings’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    strsplit\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘VariantAnnotation’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:base’:\n",
      "\n",
      "    tabulate\n",
      "\n",
      "\n",
      "Loading required package: BiocFileCache\n",
      "\n",
      "Loading required package: dbplyr\n",
      "\n",
      "\n",
      "Attaching package: ‘AnnotationHub’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    cache\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:rtracklayer’:\n",
      "\n",
      "    hubUrl\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "if (!require(\"BiocManager\", quietly = TRUE))\n",
    "    install.packages(\"BiocManager\")\n",
    "\n",
    "BiocManager::install(version = \"3.15\")\n",
    "BiocManager::install()\n",
    "BiocManager::install(\"AnnotationHub\")\n",
    "BiocManager::install(\"StructuralVariantAnnotation\")\n",
    "library(StructuralVariantAnnotation)\n",
    "library(AnnotationHub)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "ad7a73e0",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[1] ‘3.15’"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "version()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "41e9ad66",
   "metadata": {},
   "source": [
    "Downloading genes annotation for hg19"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c89907b1",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "library(AnnotationHub)\n",
    "ah <- AnnotationHub()\n",
    "ah <- subset(ah, species == \"Homo sapiens\")\n",
    "ah"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e2d6f762",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Installing package into ‘/usr/local/lib/R/site-library’\n",
      "(as ‘lib’ is unspecified)\n",
      "\n",
      "Warning message:\n",
      "“display,Hub-method is deprecated.”\n",
      "\n",
      "Listening on http://127.0.0.1:7715\n",
      "\n"
     ]
    }
   ],
   "source": [
    "install.packages(\"shiny\")\n",
    "hist <- display(ah)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e4337d29",
   "metadata": {},
   "outputs": [],
   "source": [
    "Human_gff = query(ah, c(\"Gencode\", \"gff\", \"human\"))"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.0"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
